### PR TITLE
Add another patch for numpy 2.0 (v1.84.x branch)

### PR DESCRIPTION
**Pull request by bkpoon**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

This fixes an issue that came up starting with numpy 1.21 (segmentation faults in certain cases where a C++ array was exposed to Python), but was patched in numpy. numpy 2 removed the fix so the solution is to fix Boost. More details are described in this issue (https://github.com/boostorg/python/issues/376). The upstream PR in Boost is here (https://github.com/boostorg/python/pull/443).

### CODE REVIEW

**Comment by bkpoon** [comment]
@conda-forge-admin please rerender

File: recipe/patches/0004-Support-numpy-2.0.patch [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+-            reinterpret_cast<PyArray_Descr*>(b.ptr())
+-        );
+-    }
+-#else
+-    return PyArray_EquivTypes(
+-        reinterpret_cast<PyArray_Descr*>(a.ptr()),
+-        reinterpret_cast<PyArray_Descr*>(b.ptr())
+-    );
+-#endif
++  return a == b;
```
Comment (h-vetinari) [inline recipe/patches/0004-Support-numpy-2.0.patch:35]: Does this work for all numpy versions >=1.19? If not, we'll need to add a corresponding lower bound to the `run:` requirements of `libboost-python`.

Comment (bkpoon) [reply recipe/patches/0004-Support-numpy-2.0.patch:35]: I have not tested other `numpy` versions, so setting the lower bound may be the safest option.

Comment (bkpoon) [reply recipe/patches/0004-Support-numpy-2.0.patch:35]: Is there a way to set the minimum if `numpy` is an optional dependency? Alternatively, I can test my code for `numpy` versions less than 2.0 to check if the fix works generally.

Comment (xhochy) [reply recipe/patches/0004-Support-numpy-2.0.patch:35]: You can add a `run_constraints` constraint.

**Comment by bkpoon** [comment]
I tested the patched Boost libraries with the code that I maintain with numpy versions 1.19 to 2.0, and the tests pass with the patched libraries. We have numerical arrays exposed to Python using Boost and we get segmentation faults without the fix ([numpy 2 migrator for cctbx-base package](https://github.com/conda-forge/cctbx-base-feedstock/pull/93))

I ran the tests with GitHub Actions and tested linux, Windows, and macOS (Intel and Apple Silicon). The patched packages for Boost and cctbx-base are on my testing channel (https://anaconda.org/cctbx-dev).

| numpy | Python | Test results |
| --- | --- | --- |
| 1.19 | 3.9 | https://github.com/cctbx/cctbx/actions/runs/10015271694 |
| 1.20 | 3.9 | https://github.com/cctbx/cctbx/actions/runs/10015685246 |
| 1.21 | 3.9, 3.10 | https://github.com/cctbx/cctbx/actions/runs/10015783111 |
| 1.22 | 3.9, 3.10 | https://github.com/cctbx/cctbx/actions/runs/10016155821 |
| 1.23 | 3.9, 3.10, 3.11 | https://github.com/cctbx/cctbx/actions/runs/10016930571 |
| 1.24 | 3.9, 3.10, 3.11 | https://github.com/cctbx/cctbx/actions/runs/10017112169 |
| 1.25 | 3.9, 3.10, 3.11 | https://github.com/cctbx/cctbx/actions/runs/10020687839 |
| 1.26 | 3.9, 3.10, 3.11, 3.12 | https://github.com/cctbx/cctbx/actions/runs/10021771130 |
| 2.0 | 3.9, 3.10, 3.11, 3.12 | https://github.com/cctbx/cctbx/actions/runs/10023084470 |

There is a `Test environment info` step that lists the packages in the testing environment and a `Run subset of tests` step for the actual tests that are run.

My concern with `run_constrained` is that there is probably code that is wrapped with Boost.Python, but does not go through this particular code path. If they have `numpy` in their environment, it would be updated unnecessarily. Considering that there have not been any other issues raised about this seems to indicate that this code path is not too common.

I can add the `run_constrained` option if always updating `numpy` is okay. Thoughts? Thanks!

**Comment by jakirkham** [comment]
What are the next steps here?

**Comment by h-vetinari** [comment]
> What are the next steps here?

Rebase, or at least fix the conflict.

> I can add the `run_constrained` option if always updating `numpy` is okay. Thoughts?

Thanks for the extensive testing. IIUC, things work with numpy 1.19 (which is the absolute lowest bound that any package produced here would require anyway), so then we don't need to add anything else here.

**Comment by jakirkham** [comment]
@conda-forge-admin , please re-render

**Comment by jakirkham** [comment]
That should fix the conflict

File: recipe/meta.yaml [hunk 2]
```diff
@@ -24,12 +24,14 @@ source:
...
     - patches/0001-Add-default-value-for-cxx-and-cxxflags-options-for-t.patch
     # backport https://github.com/boostorg/locale/pull/212
     - patches/0002-Reimplement-string_set-as-any_string.patch
     # backport https://github.com/boostorg/python/pull/432
     - patches/0003-Support-numpy-2.0.0b1.patch
+    # backport https://github.com/boostorg/python/pull/443
+    - patches/0004-Support-numpy-2.0.patch
 
 build:
   number: 4
```
Comment (h-vetinari) [inline recipe/meta.yaml:37]:
```suggestion
  number: 5
```

**Comment by jakirkham** [comment]
Anything else?

**Comment by bkpoon** [comment]
Checking in to see if anything else needs to be done. Thanks!

**Review by h-vetinari — APPROVED** [review]

**Comment by h-vetinari** [comment]
Thank you!

**Comment by bkpoon** [comment]
Thanks! Feel free to ping me if there are any issues.

**Pull request merged**